### ISSUE TITLE
Support Base URL

### ISSUE THREAD

**Issue by hailg** [auth0/angular2-jwt#31]
Hi, I think it would be nice to support baseURL. We can add 1 more field to AuthConfig and add that base URL into the requestHelper method.

**Comment by chenkie** [comment]
I think that would add some convenience. One thing I'd worry about is that we should probably try to keep the API similar to `Http`, which doesn't allow us to set a base url. Would it work just as well to set a constant for the base url so that requests look the same for both?

**Comment by escardin** [comment]
@hailg I think it's out of scope for AuthHttp, and will create problems if you ever want to access more than one api. That sort of thing is something you should write your own wrapper around AuthHttp and Http for. It's really quite easy, and gives you substantially more control than adding an option to AuthHttp.

For example, you could setup general error logging, or redirect to the login page if unauthenticated, or transform all your responses into objects from json.

```typescript
export class HttpWrapper{
constructor(private apiRoot:ApiRoot,private authHttp:AuthHttp) {}
 get(url: string, options?: RequestOptionsArgs): any{
        return this.authHttp.get(this.apiRoot.root+url,x).do(,logerror).map(r=>r.json());
    }
```

If you want to so something more complicated, you can do like AuthHttp does and route everything through request. It's a little more complex, but it means you can have behaviour for all requests. It really doesn't even need to be as comlex as AuthHttp's request is because you're in control of what you send it.

**Comment by hailg** [comment]
@escardin, actually, I'm using a wrapper just like what you said. But there are some reasons I think this should be integrated with this project:

1. Base URL can be an optional parameter, so if you need to call multiple API (suppose to be different domains), then you can ignore it when setup. But honestly, I think for each API host I will create different AuthHttp client to deal with different token types. Of course, if all API hosts belong to you, you can share 1 token but we are talking about general cases.

2. Well, I think you gave a valid point about sophisticated use cases then we should definitely create a wrapper. But hmm, I'm talking only about Base URL, it's very easy to add it when with the current source code.

**Comment by escardin** [comment]
If you think it's that useful, it should go into Http, not AuthHttp. AuthHttp is about Authentication, and adding misc features beyond that scope I believe does a net disservice to the user should they move on from this library.

Maybe I'm just too much of a unix do one thing type person.

You may only be talking about one thing, but next someone is going to ask for a general logging callback, and it's just one thing, and it's easier to add than a base url so...

I think that features not directly about authentication should be elsewhere. Even stuff like retrying when you get a 401 Unauthorized response is out of scope.

**Comment by chenkie** [comment]
I agree with @escardin. The goals of angular2-jwt are more to be a thin wrapper for authenticated requests with a JWT helper. :)

**Comment by hailg** [comment]
Okay, I agree that this should be added to Http instead. :)

**Comment by gustavolira** [comment]
Hi I'm having trouble to use AuthHttp wrapper, please someone could help me with this?
I did:
```typescript
@Injectable()
export class ExtAuthHttp {
  baseURL: string = 'http://localhost:8080/issuer';
  constructor(public authHttp:AuthHttp) {
  }

  get(url:string, options?:RequestOptionsArgs):any {
    console.log(this.authHttp.get(this.baseURL+url, options));
    return this.authHttp.get(this.baseURL+url, options);
  }
}
```

I want to catch 401 error and convert the result to json, but when I do for example 
` this.authHttp.get(this.baseURL+url, options).map(r=>r.json());`
I receive this error message `TypeError: this.authHttp.get(...).map is not a function in [null]`
authHttp is not null if I just return `this.authHttp.get(this.baseURL+url, options);` it works

**Comment by escardin** [comment]
@gustavolira you should give this a read: https://github.com/escardin/angular2-community-faq/blob/master/rxjs_operators.md

tl;dr you need to import your rxjs operators, i.e `import 'rxjs/add/operator/map';`

If you want to catch 401's anything that isn't a 2XX status code will go down the error path, and you will need to `.catch()`, `.retry()` or whatever to handle them.

Note: all of this stuff applies to ng2's Http module, which authHttp is just a wrapper for.

**Comment by escardin** [comment]
Also, you should create your own issue and not add on to a closed one.

**Comment by gustavolira** [comment]
Hi @escardin thank you for your attention, but if I put .catch() I receive the same error that is not a function

**Comment by escardin** [comment]
You need to load EVERY operator you use with rxjs individually. Read the link I sent. The reason you load them individually is there are ~80 operators, and loading them all really bloats your app, when you might use 5 at most.
